### PR TITLE
GH-43015: [Go] Reuse column compressors in IPC module

### PR DESCRIPTION
**Pull request by wkalt**
This changes the IPC writer to reuse lz4 or zstd compressors instead of instantiating new ones on each call to Write. Prior to this commit we would instantiate as many encoders per Write call as the configured number of concurrent compression workers, even though the compressors are reusable across calls.

### Are these changes tested?
I have added no tests and have not yet closely verified the patch. There's no functional change to test and I want to get agreement on the approach before solidifying it. I have not tested the lz4 change at all.
* GitHub Issue: #43015

### CODE REVIEW

**Comment by wkalt** [comment]
@zeroshade I have a ROS to Arrow transcoder I am trying to speed up and stumbled on this. I think we don't actually need the sync pool here - an alternative would be to assign each worker their own compressor and that would reduce synchronization a bit. I wrote it this way mostly to keep the change simple and isolated.

The data file I am working with contains 126238 messages that look like this:

```
{"topic":"/fix","sequence":1193398,"log_time":1479513060.001367422,"publish_time":1479513060.001367422,"data":{"header":{"seq":115877,"stamp":1479513060.000963926,"frame_id":"/imu"},"status":{"status":0,"service":0},"latitude":37.39954376220703,"longitude":-122.10643005371094,"altitude":-8.820882797241211,"position_covariance":[0,0,0,0,0,0,0,0,0],"position_covariance_type":0}}
```

My benchmark converts this file from MCAP to Arrow format. Baseline looks like this:

```
[~/projects/dp3/util/schema] (task/expression-evaluation) $ go test -benchmem -run=^$ -count 5 -bench ^Bench -cpuprofile cpu.prof -memprofile mem.prof                                                                                                                                                                         
goos: linux                                                                                                                                                                                                                                                                                                                    
goarch: amd64                                                                                                                                                  
pkg: github.com/wkalt/dp3/util/schema                                                                                                                          
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz                                                                                                                                                                                                                                                                                  
BenchmarkArrowConversion-8             2         623131080 ns/op        2300084172 B/op   615378 allocs/op                                                     
BenchmarkArrowConversion-8             2         639044632 ns/op        2300142164 B/op   615470 allocs/op                                                                                                                                                                                                                     
BenchmarkArrowConversion-8             2         684075134 ns/op        2300133412 B/op   615438 allocs/op                                                                                                                                                                                                                     
BenchmarkArrowConversion-8             2         625924289 ns/op        2300128356 B/op   615453 allocs/op                                                                                                                                                                                                                     
BenchmarkArrowConversion-8             2         612030702 ns/op        2300119648 B/op   615439 allocs/op
```

With this patch, like this:

```
(pprof) [~/projects/dp3/util/schema] (task/expression-evaluation) $ go test -benchmem -run=^$ -count 5 -bench ^Bench -cpuprofile cpu.prof -memprofile mem.prof
goos: linux
goarch: amd64
pkg: github.com/wkalt/dp3/util/schema
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkArrowConversion-8             3         390956287 ns/op        523875962 B/op    589463 allocs/op
BenchmarkArrowConversion-8             3         396933067 ns/op        586279402 B/op    590974 allocs/op
BenchmarkArrowConversion-8             3         409683638 ns/op        549911458 B/op    590131 allocs/op
BenchmarkArrowConversion-8             3         381790188 ns/op        508226698 B/op    589033 allocs/op
BenchmarkArrowConversion-8             3         389944825 ns/op        575889501 B/op    590775 allocs/op
PASS
ok      github.com/wkalt/dp3/util/schema        12.067s
```

so, roughly 30-35% reduction in time/op, and ~75% reduction in bytes allocated. I am using a batch size of 1K records into my record batches.  I can't split my benchmark out of my project very easily but would expect this effect to be pretty easy to reproduce. Let me know what you think.

**Comment by wkalt** [comment]
After sleeping on this I am thinking,
* The sync pools are too heavyweight and we don't want unused variables for users that aren't using one or the other compression. However we do want reuse. The Writer should maintain a slice of compressors with the same length as the number of compression workers, and the workers can use the corresponding compressor. The compressors can be lazily established with the previous version of the function so they aren't allocated if they aren't used.
* There are still reasons a user may want something like this sync pool approach. To support that I think a direction to consider would be to make the compressor interface public and allow the user to supply a factory function. This would enable both reuse of a compressor across multiple threads, and custom configuration of zstd/lz4 compressors as well as usage of alternative implementations.

**Comment by wkalt** [comment]
Now refactored according to the first point above. Performance results are comparable to the sync pool approach:

```
[~/projects/dp3/util/schema] (task/expression-evaluation) $ go test -benchmem -run=^$ -count 5 -bench ^Bench -cpuprofile cpu.prof -memprofile mem.prof
goos: linux
goarch: amd64
pkg: github.com/wkalt/dp3/util/schema
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkArrowConversion-8             3         442864629 ns/op        374034618 B/op    838037 allocs/op
BenchmarkArrowConversion-8             3         420247966 ns/op        374016525 B/op    838024 allocs/op
BenchmarkArrowConversion-8             3         414452686 ns/op        373981138 B/op    838025 allocs/op
BenchmarkArrowConversion-8             3         420451169 ns/op        374095818 B/op    838029 allocs/op
BenchmarkArrowConversion-8             3         410123887 ns/op        374109312 B/op    838025 allocs/op
```

File: go/arrow/ipc/ipc.go [hunk 1]
```diff
@@ -135,6 +136,9 @@ func WithZstd() Option {
 // parallelization. Default is 0.
```
Comment (zeroshade) [inline go/arrow/ipc/ipc.go:136]: Should update the comment that the default is now 1

**Review by zeroshade — COMMENTED** [review]
This looks good to me overall, thanks much for this! I'm glad to see the benefits and improvements. I just have a single nit pick :smile:

File: go/arrow/ipc/ipc.go (hunk 1 above)
Comment (wkalt) [reply go/arrow/ipc/ipc.go]: Thanks, updated.

**Review by zeroshade — APPROVED** [review]

**Pull request merged**